src/EditorFeatures/CSharp/Highlighting/KeywordHighlighters/AsyncAwaitHighlighter.cs

```diff
@@ -26,93 +30,90 @@ protected override bool IsHighlightableNode(SyntaxNode node)
-                        var asyncModifier = localFunction.Modifiers.FirstOrDefault(m => m.Kind() == SyntaxKind.AsyncKeyword);
-                        if (asyncModifier.Kind() != SyntaxKind.None)
-                        {
-                            spans.Add(asyncModifier.Span);
-                        }
-                        break;
-                    }
-                case AnonymousFunctionExpressionSyntax anonymousFunction:
-                    if (anonymousFunction.AsyncKeyword.Kind() == SyntaxKind.AsyncKeyword)
-                    {
-                        spans.Add(anonymousFunction.AsyncKeyword.Span);
-                    }
-                    break;
-
-                case AwaitExpressionSyntax awaitExpression:
-                    // Note if there is already a highlight for the previous token, merge it
-                    // with this span. That way, we highlight nested awaits with a single span.
-                    var handled = false;
-                    var awaitToken = awaitExpression.AwaitKeyword;
-                    var previousToken = awaitToken.GetPreviousToken();
-                    if (!previousToken.Span.IsEmpty)
-                    {
-                        var index = spans.FindIndex(s => s.Contains(previousToken.Span));
-                        if (index >= 0)
-                        {
-                            var span = spans[index];
-                            spans[index] = TextSpan.FromBounds(span.Start, awaitToken.Span.End);
-                            handled = true;
-                        }
-                    }
+            using var pooledObject = s_stackPool.GetPooledObject();
 
-                    if (!handled)
-                    {
-                        spans.Add(awaitToken.Span);
-                    }
-                    break;
+            var stack = pooledObject.Object;
+            stack.Push(node);
 
-                case UsingStatementSyntax usingStatement:
-                    if (usingStatement.AwaitKeyword.Kind() == SyntaxKind.AwaitKeyword)
-                    {
-                        spans.Add(usingStatement.AwaitKeyword.Span);
-                    }
-                    break;
+            while (stack.Count > 0)
+            {
+                var current = stack.Pop();
+                yield return current;
 
-                case LocalDeclarationStatementSyntax localDeclaration:
-                    if (localDeclaration.AwaitKeyword.Kind() == SyntaxKind.AwaitKeyword && localDeclaration.UsingKeyword.Kind() == SyntaxKind.UsingKeyword)
+                foreach (var child in current.ChildNodes())
+                {
+                    // Only process children if they're not the start of another construct
+                    // that async/await would be related to.
+                    if (!child.IsReturnableConstruct())
                     {
-                        spans.Add(localDeclaration.AwaitKeyword.Span);
+                        stack.Push(child);
```

Comment:
💭 I'm worried how big this stack can get in practice. Will review with @dpoeschl

Reply:
an trees we walk down could potentially be unbounded in length.